pkg/certificates/authorizations/db.go

```diff
@@ -0,0 +1,271 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package authorizations
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/internal/dbutil"
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/peertls/extensions"
+	"storj.io/storj/storage"
+	"storj.io/storj/storage/boltdb"
+	"storj.io/storj/storage/redis"
+)
+
+// DB stores authorizations which may be claimed in exchange for a
+// certificate signature.
+type DB struct {
+	db storage.KeyValueStore
+}
+
+// Config is the authorization db config.
+type Config struct {
+	DBURL     string `default:"bolt://$CONFDIR/authorizations.db" help:"url to the certificate signing authorization database"`
+	Overwrite bool   `default:"false" help:"if true, overwrites config AND authorization db is truncated" setup:"true"`
+}
+
+// NewDBFromCfg creates and/or opens the authorization database specified by the config.
+func NewDBFromCfg(config Config) (*DB, error) {
+	return NewDB(config.DBURL, config.Overwrite)
+}
+
+// NewDB creates and/or opens the authorization database.
+func NewDB(dbURL string, overwrite bool) (*DB, error) {
+	driver, source, err := dbutil.SplitConnstr(dbURL)
+	if err != nil {
+		return nil, extensions.ErrRevocationDB.Wrap(err)
+	}
+
+	authDB := new(DB)
+	switch driver {
+	case "bolt":
+		_, err := os.Stat(source)
+		if overwrite && err == nil {
+			if err := os.Remove(source); err != nil {
+				return nil, err
+			}
+		}
+
+		authDB.db, err = boltdb.New(source, AuthorizationsBucket)
+		if err != nil {
+			return nil, ErrAuthorizationDB.Wrap(err)
+		}
+	case "redis":
+		redisClient, err := redis.NewClientFrom(dbURL)
+		if err != nil {
+			return nil, ErrAuthorizationDB.Wrap(err)
+		}
+
+		if overwrite {
+			if err := redisClient.FlushDB(); err != nil {
+				return nil, err
+			}
+		}
+
+		authDB.db = redisClient
+	default:
+		return nil, ErrAuthorizationDB.New("database scheme not supported: %s", driver)
+	}
+
+	return authDB, nil
+}
+
+// Close closes the authorization database's underlying store.
+func (authDB *DB) Close() error {
+	return ErrAuthorizationDB.Wrap(authDB.db.Close())
+}
+
+// Create creates a new authorization and adds it to the authorization database.
+func (authDB *DB) Create(ctx context.Context, userID string, count int) (_ Authorizations, err error) {
+	defer mon.Task()(&ctx)(&err)
+	if len(userID) == 0 {
+		return nil, ErrAuthorizationDB.New("userID cannot be empty")
+	}
+	if count < 1 {
+		return nil, ErrAuthorizationCount
+	}
+
+	var (
+		newAuths Authorizations
+		authErrs errs.Group
+	)
+	for i := 0; i < count; i++ {
+		auth, err := NewAuthorization(userID)
+		if err != nil {
+			authErrs.Add(err)
+			continue
+		}
+		newAuths = append(newAuths, auth)
+	}
+	if err := authErrs.Err(); err != nil {
```

Comment:
I think this can be just `if authErrs.Err() != nil {`

Reply:
I think that's used in this way for not calling `authErrs.Err()` twice but the outcome is the same.

Reply:
oh you might be right